ktesting/testinglogger.go

```diff
@@ -227,35 +247,85 @@ func (b *buffer) Data() Log {
 	return b.log.DeepCopy()
 }
 
+// tloggerShared holds values that are the same for all instances.
+type tloggerShared struct {
+	// mutex protects access to t.
+	mutex sync.Mutex
+
+	// t gets cleared when the test is completed.
+	t TL
+
+	testName  string
+	config    *Config
+	buffer    buffer
+	callDepth int
+}
+
+func (ls *tloggerShared) stop() {
+	ls.mutex.Lock()
+	defer ls.mutex.Unlock()
+	ls.t = nil
+}
+
+// tlogger is the actual LogSink implementation.
 type tlogger struct {
-	t      TL
+	shared *tloggerShared
 	prefix string
 	values []interface{}
-	config *Config
-	buffer *buffer
 }
 
-func (l *tlogger) Init(info logr.RuntimeInfo) {
+func (l tlogger) fallbackLogger() logr.Logger {
+	logger := klog.Background().WithValues(l.values...).WithName(l.shared.testName + " leaked goroutine")
+	if l.prefix != "" {
+		logger = logger.WithName(l.prefix)
+	}
+	// Skip direct caller (= Error or Info) plus the logr wrapper.
+	logger = logger.WithCallDepth(l.shared.callDepth + 1)
+	return logger
+}
+
+func (l tlogger) Init(info logr.RuntimeInfo) {
+	l.shared.callDepth = info.CallDepth
 }
 
-func (l *tlogger) GetCallStackHelper() func() {
-	return l.t.Helper
+func (l tlogger) GetCallStackHelper() func() {
+	l.shared.mutex.Lock()
+	defer l.shared.mutex.Unlock()
+	if l.shared.t == nil {
+		return func() {}
+	}
+
+	return l.shared.t.Helper
 }
 
-func (l *tlogger) Info(level int, msg string, kvList ...interface{}) {
-	l.t.Helper()
+func (l tlogger) Info(level int, msg string, kvList ...interface{}) {
+	l.shared.mutex.Lock()
+	defer l.shared.mutex.Unlock()
+	if l.shared.t == nil {
+		l.fallbackLogger().V(level).Info(msg, kvList...)
```

Comment:
Would it be better to log this message without using the `V(level)` switch? Would that help catch some of the issue with tests where things are still getting written to log after test is done?

Reply:
That's an interesting idea. I'm just worried that this might increase the log volume of tests which currently have no way of cleaning up goroutines.

What if we instead print one warning per test logger the first time that a leaked goroutine is detected? That warning then can provide information about that goroutine, which is more than a developer gets from the prefix.

Reply:
I added:
```
        testinglogger_test.go:195] "TestStop/Sub leaked goroutine: WARNING: test kept at least one goroutine running after test completion" callstack=<
        	goroutine 35 [running]:
        	k8s.io/klog/v2/internal/dbg.Stacks(0x0)
        		/nvme/gopath/src/k8s.io/klog/internal/dbg/dbg.go:34 +0x8a
        	k8s.io/klog/v2/ktesting.tlogger.fallbackLogger({0xc000160780, {0x0, 0x0}, {0x0, 0x0, 0x0}})
        		/nvme/gopath/src/k8s.io/klog/ktesting/testinglogger.go:292 +0x232
        	k8s.io/klog/v2/ktesting.tlogger.Info({0xc000160780, {0x0, 0x0}, {0x0, 0x0, 0x0}}, 0x0, {0x5454f1, 0x13}, {0x0, ...})
        		/nvme/gopath/src/k8s.io/klog/ktesting/testinglogger.go:316 +0x28a
        	github.com/go-logr/logr.Logger.Info({{0x5735f8?, 0xc00011d020?}, 0x0?}, {0x5454f1, 0x13}, {0x0, 0x0, 0x0})
        		/nvme/gopath/pkg/mod/github.com/go-logr/logr@v1.2.0/logr.go:249 +0xd0
        	k8s.io/klog/v2/ktesting_test.TestStop.func1.1()
        		/nvme/gopath/src/k8s.io/klog/ktesting/testinglogger_test.go:195 +0xd0
        	created by k8s.io/klog/v2/ktesting_test.TestStop.func1
        		/nvme/gopath/src/k8s.io/klog/ktesting/testinglogger_test.go:186 +0xf8
         >
```

This could be made a bit nicer by skipping frames, but I want to keep the code simple, so I am just using the same helper as for goroutine dumps in klog (= wrapper around runtime.Stacks), except that I am only dumping one goroutine.

Dumping only one is intentional: "go test" itself uses goroutines and the test suite might run other tests in parallel.

Reply:
Makes sense. This already give us good enough good information to start digging into failures better.

Reply:
Please mark this as resolved